codegen/src/errors.rs

```diff
@@ -22,5 +22,8 @@ error_chain! {
             display("cannot convert protobuf import into module import: {}", desc)
             cause("import definition might be invalid, some characters may not be supported")
         }
+        EmptyRead {
+            description("No messages or enums were read; either there was no input or there were only unsupported structures")
```

Comment:
Sorry I did not see this comment!

I'm not exactly sure why you can write `cause(...)` with `error_chain` but I understood that `cause`s are only for wrapped errors. You could for example have a call a function that returns `std::io::Error` and use `ResultExt::chain_err` to wrap it into more understandable, only for printing purposes though. `error_chain` does not support introspection with chained errors using `match`. I think line 23 works because `&'static str` is acceptable as an inner `Error + 'static + Send + Sync` using some magic but I do not think this is the right way to use it as a sort of "more details". Though, I cannot imagine why it would be there in the macro if it wasn't for this.

My understanding comes from [ResultExt::chain_err](https://docs.rs/error-chain/0.8.1/error_chain/example_generated/trait.ResultExt.html#tymethod.chain_err), [ErrorChainIter](https://docs.rs/error-chain/0.8.1/error_chain/struct.ErrorChainIter.html) and the original [Error::cause](https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.cause) but I find the `error_chain` docs a bit hard to follow at some points.

Did some more digging on this and found that `error_chain` uses [`quick_error`](https://github.com/tailhook/quick-error) underneath (macro calling macro, calling macro, ...) and `quick_error`s README.md has an example where `cause(err)` is used to make `Error::cause` return the value inside the ErrorKind enum variant. So not meant for details.

Implementing the "more details" could probably be done with something like:

```
trait DetailedError {
    fn detailed_description(&self) -> Option<&str>;
}

impl DetailedError for errors::Error {
    fn detailed_description(&self) -> Option<&str> {
        match *self {
          (errors::ErrorKind::EmptyRead, _) => Some("Either there was no input or there were only unsupported structures"),
          (errors::ErrorKind::InvalidImport, _) => Some("..."),
          _ => None
    }
}

/// Newtype for formatting detailed errors, as we cannot override
/// their Display implementation generated by `error_chain`
struct DetailedDisplay<E: std::error::Error + DetailedError>(E);

impl std::fmt::Display for DetailedDisplay<E: std::error::Error + DetailedError> {
    fn fmt(...) -> std::fmt::Result {
        // like normal display for Errors, but with the possible details
    }
}
```

But above is just one way this could be implemented. If this is really wanted, I guess first step could be to investigate how do `rustc` errors work as they have all kinds of hints and suggestions as details.